contrib/peak_ewma/load_balancing_policies/source/peak_ewma_lb.cc

```diff
@@ -0,0 +1,338 @@
+        auto it = all_host_stats_.find(host);
+        if (it == all_host_stats_.end()) {
+          all_host_stats_[host] = observability_.createHostStats(host);
+          it = all_host_stats_.find(host);
+        }
+
+        // Update stats for observability
+        if (it != all_host_stats_.end()) {
+          it->second->setEwmaRttStat(ewma_rtt);
+          it->second->setActiveRequestsStat(active_requests);
+          it->second->setComputedCostStat(cost);
+        }
+
+        // Host processing complete
+      }
+    }
+  }
+
+  // Aggregation cycle complete
+}
+
+size_t PeakEwmaLoadBalancer::calculateNewSampleCount(size_t last_processed, size_t current_write,
+                                                     size_t max_samples) {
+  if (last_processed == current_write) {
+    return 0;
+  }
+
+  if (current_write >= last_processed) {
+    return current_write - last_processed;
+  } else {
+    // Write index wrapped around
+    return (max_samples - last_processed) + current_write;
+  }
+}
+
+double PeakEwmaLoadBalancer::calculateTimeBasedAlpha(uint64_t current_time_ns,
+                                                     uint64_t sample_time_ns) {
+  int64_t time_delta_ns = static_cast<int64_t>(current_time_ns - sample_time_ns);
+  if (time_delta_ns <= 0) {
+    return 1.0; // Use full weight for future/concurrent samples
+  }
+
+  // Time-based exponential decay: α = 1 - e^(-Δt/τ)
+  double time_delta_s = time_delta_ns / 1000000000.0;
+  double tau_s = tau_nanos_ / 1000000000.0;
+  double alpha = 1.0 - std::exp(-time_delta_s / tau_s);
+
+  // Clamp alpha to reasonable bounds
+  return std::min(1.0, std::max(0.0, alpha));
+}
+
+double PeakEwmaLoadBalancer::updateEwmaWithSample(double current_ewma, double new_rtt_ms,
+                                                  double alpha) {
+  if (current_ewma == 0.0) {
+    // First sample - initialize EWMA
+    return new_rtt_ms;
+  }
+
+  // EWMA update: new_ewma = α × new_rtt + (1-α) × old_ewma
+  return alpha * new_rtt_ms + (1.0 - alpha) * current_ewma;
+}
+
+void PeakEwmaLoadBalancer::processHostSamples(Upstream::HostConstSharedPtr /* host */,
+                                              PeakEwmaHostLbPolicyData* data) {
+  if (!data)
+    return;
+
+  // Get the range of new samples to process (atomic ring buffer)
+  auto [last_processed, current_write] = data->getNewSampleRange();
+
+  size_t num_new_samples =
+      calculateNewSampleCount(last_processed, current_write, data->max_samples_);
+  if (num_new_samples == 0)
+    return;
+
+  // Get current EWMA state
+  double current_ewma = data->getEwmaRtt();
+  uint64_t current_time_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
+                                 time_source_.monotonicTime().time_since_epoch())
+                                 .count();
+
+  // Process all new samples in chronological order
+  size_t processed_index = last_processed;
+  for (size_t i = 0; i < num_new_samples; ++i) {
```

Comment:
How do you handle `num_new_samples` being larger than the max samples? I'm either missing something or this loop will process some samples more than once under high load.